sql/settype.go

```diff
@@ -0,0 +1,359 @@
+package sql
+
+import (
+	"fmt"
+	"math/bits"
+	"sort"
+	"strings"
+
+	"gopkg.in/src-d/go-errors.v1"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/proto/query"
+)
+
+var (
+	ErrConvertingToSet = errors.NewKind("value %v is not valid for this Set")
+	ErrDuplicateEntrySet = errors.NewKind("duplicate entry: %v")
+	ErrInvalidSetValue = errors.NewKind("value %v was not found in the set")
+)
+
+// Comments with three slashes were taken directly from the linked documentation.
+
+// Represents the SET type.
+// https://dev.mysql.com/doc/refman/8.0/en/set.html
+type SetType interface {
+	Type
+	CharacterSet() CharacterSet
+	Collation() Collation
+	ConvertBitArrayToString(bitArray []uint64) (string, error)
```

Comment:
These two convert methods should either be private or not on the interface in general. The fact these are stored as a bitfield is an implementation detail that shouldn't be exposed.